### ISSUE TITLE
[CustomScript] Extension Does Not Support Hot Storage

### ISSUE THREAD

**Issue by boumenot** [Azure/azure-linux-extensions#161]
v1.5.2.0.

The extension failed to fetch a script from hot storage, and failed with "The value for one of the HTTP headers is not in the correct format."

The contract when communicating with blob storage should be consistent when communicating between the various flavors, but there appears to be a break down with hot storage.  It's unclear where the error is, and more investigation is needed.

**Comment by paulmey** [comment]
Looks like CustomScript has a copy of the blob client in it's tree. I replaced it with a newer version on a runnning VM, which needed quite some changes and left me with other warnings, but eventually it worked.
Pulling in the azure-storage package and anything needed to do proper SSL requests (pyOpenSSL, probably doesn't work everywhere anyway, since it compiles against shared openssl) seems like overkill to me. IMHO, the full blob client is already overkill and all we should only 'borrow' is the code that calculates the request signature. After all, that's all we need for the HTTP get to succeed...

What do you think @boumenot ?

**Comment by jasonzio** [comment]
So, every time there's a new feature or capability added to storage, we'll have to re-address the borrowed code... ? Or, just use the published SDK and update to the appropriate version...

Seems the latter choice is wisest. It increases the official dependency set for the extension, but it also acknowledges that the dependency has existed all along.

**Comment by boumenot** [comment]
Agreed.  I am not against embedding the SDK, but I don't like the idea of having a copy that's checked in.   I think it is too much to expect our users to install Python libraries to satisfy an extension, and it may create chicken and egg issues.

## Ideas

1. Boost's [bcp][1] is a neat idea, and something for Python may already exist. **[Investigate]**
1. Lock onto a version of the SDK using pip's requirements.txt, and as part off the build.
 * Download and extract the specified SDK version.
 * Copy the necessary file(s) from the SDK into the CustomScript's directory.
1. I like the way the Docker extension handles this, which is by not.  It ships a complete binary blob.  I think this is a future direction, but it will not satisfy our immediate needs.

[1]: http://www.boost.org/doc/libs/1_61_0/tools/bcp/doc/html/index.html

**Comment by ahmetalpbalkan** [comment]
IMO most viable solution right now is to just have a `common.py` and copy it around to all extensions. It is mostly static anyway.

If you look at my list at https://github.com/Azure/WALinuxAgent/issues/223#issuecomment-224713170, it should not be that hard to apply a refactoring and extract common functionality into a single file.

**Comment by tomconte** [comment]
Hi there, I have a partner who is receiving the same error message, but using Zone-Redundant Storage (ZRS). Is it the same cause (x-ms-version too old)? If yes could we update the README? Thanks!

**Comment by tomconte** [comment]
Follow-up question: as a quick fix, could we just update the x-ms-version to e.g. 2014-06-01 so we support the newer storage account types? I think this is the only required change since the new storage accounts are API-compatible with the general purpose ones.

**Comment by boumenot** [comment]
I would be leery of just changing the date to enable support.  We should upgrade the SDK version to get the necessary support.

@tomconte There is a newer version of [CustomScript][1] that we are experimenting with to address some of the issues that we have this version (this issue is a good example).  Is this something that you could use to unblock you now, while we integrate a newer version of the library?

At some point we expect these two versions to converge.

[1]: https://github.com/Azure/custom-script-extension-linux

**Comment by ahmetalpbalkan** [comment]
@boumenot have you looked at what's blocking the existing version from working with Hot Storage? We didn't do anything special so I bet it's not going to work out of the box.

**Comment by boumenot** [comment]
No, I have not investigated.  I was commenting that I would not bump date, and redeploy CustomScriptForLinux.  I would be willing to integrate the latest SDK version, and see if that addresses the issue.

**Comment by tomconte** [comment]
@boumenot I'd be happy to try the new version, but how do I use it from an ARM template instead of the default CustomScriptForLinux extension?

**Comment by boumenot** [comment]
It is meant to be a drop in replacement.  Change the name to Microsoft.Azure.Extensions.CustomScript, and set the version to 2.0.
